### DISCUSSION TITLE
multiple instances of nitro (microservices) in nuxt - possible ?

### DISCUSSION THREAD

**Discussion by orenmizr** [nuxt/nuxt#23102] [Questions] [answered]
hi, i have a nuxt 3 setup with nitro and server/api. i want to create another server for imports queue - it is heavy in cpu so i don't want it running while nitro is serving the users so another instance of node seem the right way.

what is the nuxt/nitro elegant way of solving this issue ? is there one ? could not see it in the docs/examples

**Comment by mathix420** [comment]
I am not sure what you meant by imports queue, but if you have a CPU intensive task that does not require to be running in exactly the same instance that runs Nuxt you should use a separate instance for this task only and communicate with it from Nuxt (using an API endpoint or gRPC).

Then you can also leverage the full power of the instance running Nuxt by enabling  [Cluster Mode](https://nuxt.com/docs/getting-started/deployment#using-cluster-mode) on your app.
You can also use [PM2 in cluster mode](https://pm2.io/docs/runtime/guide/load-balancing/), allowing you to run multiple instances of node in the same instance.

If having only one instance of Nuxt is not sufficient, what you can always do is load balancing. So you can spin as much instances of nuxt you want and your load balancer will distribute the traffic evenly across them.

**Reply by orenmizr** [reply]
"...Nuxt you should use a separate instance for this task " - my question was how can i achieve separate instances ? nothing about this on nuxt docs.  i only saw examples for h3 servers on nitro site. i wanted the "nuxt way".

also, could not find a lot of into regarding nuxt and Cluster Mode.

**Reply by mathix420** [reply]
As I told you, I'm not sure what exactly your use case is with the import queues, so I can't provide you with more details.

But when I'm talking separate instances for your heavy workload I meant basic [micro service](https://cloud.google.com/learn/what-is-microservices-architecture).
Nuxt won't document how to implement this as it is totally different from the website, it is just another service of your backend.

If your use case is just to have 2 or more machines running Nuxt, you will be forced to use load balancing. This won't be documented by Nuxt as it is totally different for every deployment depending on your infrastructure. I suggest you watch some documentation or tutorials about load balancing.

If your use case is to run Nuxt in multithreaded mode (use all the available CPU cores and power from one machine), then you can use the [Cluster Mode from Nitro](https://nitro.unjs.io/deploy/node#cluster-mode) (the server engine of Nuxt).
No documentation is needed there is just a flag to enable, and that's it: your Nuxt server is multithreaded. PM2 is a similar solution, maybe more documented.

**Reply by orenmizr** [reply]
first,  thanks for responding. i really appreciate it. i should not have mentioned the queue think as it probably made it more confusing. 

i was talking about microservice architecture... when you need to perform something with different concerns from serving the browser you put it on a another service (fastify, nestjs). you can have it with routes or just listen to events from redis/kafka. 

with nuxt, which seems likes a bff, not sure how to spin another "server" because it is built under the hood. 
what i am asking exactly is: 

1. i have the frontend + backend routes with db (drizzle) + auth
2. i need another service that can take file uploads and process them. 

now, logic wise - 2 has nothing to do with 1, excepts when i am done i will enrich the db data.  
so how can i tell nuxt this another instance of a nitro server ?

**Reply by mathix420** [reply, marked as answer]
Okay, so if I understand well you just want to spin 2 or more instances of Nuxt?

There is no "Nuxt way" for this, just use a load balancer or PM2 (which is basically a local load balancer).
The two instances don't need to be aware of each other's, they will both do the same thing.
